quickwit-indexing/src/source/push_api_source.rs

```diff
@@ -0,0 +1,387 @@
+
+use super::file_source::BATCH_NUM_BYTES_THRESHOLD;
+use super::{Source, SourceContext, TypedSourceFactory};
+use crate::actors::Indexer;
+use crate::models::RawDocBatch;
+
+/// Wait time for SourceActor before pooling for new documents.
+/// TODO: Think of better way, maybe increment this (i.e wait longer) as time
+/// goes on without receiving docs.
+const PUSHAPI_POLLING_COOL_DOWN: Duration = Duration::from_secs(1);
+
+#[derive(Default, Clone, Debug, Eq, PartialEq, Serialize)]
+pub struct PushApiSourceCounters {
+    pub previous_offset: u64,
+    pub current_offset: u64,
+    pub num_docs_processed: u64,
+}
+
+pub struct PushApiSource {
+    source_id: String,
+    params: PushApiSourceParams,
+    push_api_mailbox: Mailbox<PushApiService>,
+    counters: PushApiSourceCounters,
+}
+
+impl fmt::Debug for PushApiSource {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "PushApiSource {{ source_id: {} }}", self.source_id)
+    }
+}
+
+impl PushApiSource {
+    pub async fn make(
+        source_id: String,
+        params: PushApiSourceParams,
+        push_api_mailbox: Mailbox<PushApiService>,
+        checkpoint: SourceCheckpoint,
+    ) -> anyhow::Result<Self> {
+        let partition_id = PartitionId::from(params.index_id.clone());
+        let offset = if let Some(Position::Offset(offset_str)) =
+            checkpoint.position_for_partition(&partition_id).cloned()
+        {
+            offset_str.parse::<u64>()?
+        } else {
+            0
+        };
+
+        // The PushApiSource is instantiated for an existing index.
+        // It's safe to create the corresponding queue if it does not exist.
+        ensure_queue_exist_for_index(&push_api_mailbox, params.index_id.clone()).await?;
+        let push_api_source = PushApiSource {
+            source_id,
+            params,
+            push_api_mailbox,
+            counters: PushApiSourceCounters {
+                previous_offset: offset,
+                current_offset: offset,
+                num_docs_processed: 0,
+            },
+        };
+        Ok(push_api_source)
+    }
+
+    fn update_counters(&mut self, current_offset: u64, num_docs: u64) {
+        self.counters.num_docs_processed += num_docs;
+        self.counters.current_offset = current_offset;
+        self.counters.previous_offset = current_offset;
+    }
+}
+
+/// Creates a queue for the index if it does not exists.
+async fn ensure_queue_exist_for_index(
+    push_api_mailbox: &Mailbox<PushApiService>,
+    index_id: String,
+) -> anyhow::Result<()> {
+    let queue_exists_req = QueueExistsRequest {
+        queue_id: index_id.clone(),
+    };
+    if push_api_mailbox
+        .ask_for_res(queue_exists_req)
+        .await
+        .map_err(|err| anyhow::anyhow!(err.to_string()))?
+    {
+        return Ok(());
+    }
+
+    let create_queue_req = CreateQueueRequest { queue_id: index_id };
+    push_api_mailbox
+        .ask_for_res(create_queue_req)
```

Comment:
there is a race condition here.

Reply:
Yes, you are right. I just tried to avoid using the error condition of CreateQueueRequest as success like you advise last time without thinking about the implication of message ordering.
Can I create a custom message for this operation? `EnsureQueueExistRequest` ?